### PR TITLE
Add support for oauth connection management in botservice

### PR DESCRIPTION
**Pull request by swagatmishra2007**
I am just checking in the feature work for now. i will publish a new extension around next week, when the associated deployment completes.

### CODE REVIEW

File: src/botservice/azext_bot/.vscode/tags [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+service_provider_parameter.py	../botservice\models\service_provider_parameter.py	1;"	kind:file	line:1
+service_provider_properties.py	../botservice\models\service_provider_properties.py	1;"	kind:file	line:1
+service_provider_response_list.py	../botservice\models\service_provider_response_list.py	1;"	kind:file	line:1
+shutil	../custom.py	/^import shutil$/;"	kind:namespace	line:8
+sku	../custom.py	/^from azext_bot.botservice.models import Bot, BotProperties, sku, BotChannel$/;"	kind:namespace	line:13
+sku.py	../botservice\models\sku.py	1;"	kind:file	line:1
+skype_channel	../botservice\models\azure_bot_service_enums.py	/^    skype_channel = "SkypeChannel"$/;"	kind:variable	line:43
+skype_channel.py	../botservice\models\skype_channel.py	1;"	kind:file	line:1
+skype_channel_properties.py	../botservice\models\skype_channel_properties.py	1;"	kind:file	line:1
+skype_create	../custom.py	/^def skype_create(client, resource_group_name, resource_name, is_disabled=None, enable_messa
```
Comment (williexu) [inline src/botservice/azext_bot/.vscode/tags]: what is this file for?

File: src/botservice/azext_bot/_help.py [hunk 2]
```diff
@@ -119,6 +119,30 @@
...
+"""
+helps['bot connection list'] = """
+    type: command
+    short-summary: Show all OAuth Connection Settings on a Bot.
+"""
+helps['bot connection delete'] = """
+    type: command
+    short-summary: Delete an OAuth Connection Setting on a Bot.
+"""
+helps['bot connection listproviders'] = """
```
Comment (williexu) [inline src/botservice/azext_bot/_help.py]: `list-providers`

File: src/botservice/azext_bot/_help.py [hunk 3]
```diff
@@ -119,6 +119,30 @@
     type: command
     short-summary: Create Slack Channel on a Bot.
 """
+helps['bot connection'] = """
+    type: command
```
Comment (williexu) [inline src/botservice/azext_bot/_help.py]: `type: group`

File: src/botservice/azext_bot/commands.py [hunk 4]
```diff
@@ -37,6 +37,13 @@ def load_command_table(self, _):
         g.custom_command('delete', 'delete_bot')
         g.generic_update_command('update', setter_name='update', setter_type=updateBotService_commandType)
 
+    with self.command_group('bot connection', botOperations_commandType) as g:
+        g.custom_command('list', 'get_connections')
+        g.custom_command('show', 'get_connection')
+        g.custom_command('create', 'create_connection')
+        g.custom_command('delete', 'delete_connection')
+        g.custom_command('listproviders', 'get_service_providers')
```
Comment (williexu) [inline src/botservice/azext_bot/commands.py]: `list-providers` is more aligned with the rest of the CLI's commands.

File: src/botservice/azext_bot/botservice/operations/bots_operations.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def create(
         # Construct URL
         url = '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.BotService/botServices/{resourceName}'
         path_format_arguments = {
-            'resourceGroupName': self._serialize.url("resource_group_name", resource_group_name, 'str'),
+            'resourceGroupName': self._serialize.url("resource_group_name", resource_group_name, 'str', max_length=64, min_length=2, pattern=r'^[a-zA-Z0-9][a-zA-Z0-9_.-]*$'),
```
Comment (williexu) [inline src/botservice/azext_bot/botservice/operations/bots_operations.py]: Is this newly generated from autorest? or are these manual changes?

File: src/botservice/azext_bot/_params.py [hunk 6]
```diff
@@ -108,3 +108,17 @@ def load_arguments(self, _):
...
+
+    with self.argument_context('bot connection create') as c:
+        c.argument('client_id', help='client id associated with the service provider setting')
+        c.argument('client_secret', help='client secret associated with the service provider setting')
+        c.argument('scopes', help='scopes associated with the service provider setting.The format depends on the service provider.')
+        c.argument('service_provider_name', options_list=['--service'], help='name of the service provider. For a list of all service providers, use az bot connection listserviceproviders')
+        c.argument('parameters', help='parameter values for Service Provider Parameters. Usage: --parameters key=value key1=value1', nargs='+')
+
+    with self.argument_context('bot connection listproviders') as c:
+        c.argument('as_raw_settings', options_list=['--asraw'], help='Output the raw json for each service provider', arg_type=get_three_state_flag())
```
Comment (williexu) [inline src/botservice/azext_bot/_params.py]: `options_list=['--as-raw']`

File: src/botservice/azext_bot/.vscode/tags (hunk 1 above)
Comment (swagatmishra2007) [reply src/botservice/azext_bot/.vscode/tags]: vs code inserts this for some reason. i will remove it

File: src/botservice/azext_bot/botservice/operations/bots_operations.py (hunk 5 above)
Comment (swagatmishra2007) [reply src/botservice/azext_bot/botservice/operations/bots_operations.py]: newly generated from autorest. i added some validation to the resource group parameter in the swagger , so it showed up here

File: src/botservice/azext_bot/.vscode/tags (hunk 1 above)
Comment (swagatmishra2007) [reply src/botservice/azext_bot/.vscode/tags]: aah generated by vscode for some reason. i will remove it. it always regenerates it every time i open the code base in vscode.

**Pull request merged**